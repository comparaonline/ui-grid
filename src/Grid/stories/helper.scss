.row {
  background: #eee;
  font-size: 11px;
  margin-bottom: 10px;
  line-height: 2rem;
  border: 1px solid #c6c6c6;
  margin-left: 0;
  margin-right: 0;
}

.row .column:nth-child(2) {
  background: #e1e1e1;
}
