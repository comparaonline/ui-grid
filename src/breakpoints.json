{
  "small": "0",
  "medium": "200px",
  "large": "1024px",
  "xlarge": "1170px",
  "xxlarge": "1306px"
}
