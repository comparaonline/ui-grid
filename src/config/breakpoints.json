{
  "xsmall": "0",
  "small": "480px",
  "medium": "600px",
  "large": "840px",
  "xlarge": "960px",
  "xxlarge": "1280px"
}
