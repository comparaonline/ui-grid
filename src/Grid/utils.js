import PropTypes from 'prop-types';
import breakpoints from '../breakpoints.json';

const bps = Object.keys(breakpoints);
const bpCamelCase = bp => `${bp.substr(0, 1).toUpperCase()}${bp.substr(1)}`;

export const bpPropClass = (bp, prop) => `${prop}On${bpCamelCase(bp)}`;

const getBooleanClassNames = (props, booleanProps) => booleanProps.reduce(
  (names, prop) => names.concat(props[prop] ? ` ${prop}` : ''), '');

const getBreakpointClassName = (props, bp) => (props[bp] ? ` ${bp}-${props[bp]}` : '');

const getBreakpointNumericClassName = (props, numericProps, bp) => numericProps.reduce(
  (names, prop) => {
    const propValue = props[bpPropClass(bp, prop)];
    return names.concat(propValue ? ` ${bp}-${prop}-${propValue}` : '');
  }, '');

const getBreakpointBooleanClassName = (props, booleanProps, bp) => booleanProps.reduce(
  (names, prop) => names.concat(props[bpPropClass(bp, prop)] ? ` ${bp}-${prop}` : ''), '');

const getBreakpointVisibilityClassName = (props, booleanProps, bp) => booleanProps.reduce(
  (names, prop) => names.concat(props[bpPropClass(bp, prop)] ? ` ${prop}-for-${bp}` : ''), '');

const generateBreakpointClassNames = (props, propNames, bp) => ''.concat(
  propNames.include ? getBreakpointClassName(props, bp) : '',
  propNames.numeric ? getBreakpointNumericClassName(props, propNames.numeric, bp) : '',
  propNames.boolean ? getBreakpointBooleanClassName(props, propNames.boolean, bp) : '',
  propNames.visibility ? getBreakpointVisibilityClassName(props, propNames.visibility, bp) : ''
);

/**
 * Generate string with classnames from props names and breakpoints
 *
 * @param {Object} props
 * @param {String Array} numericProps
 * @param {String Array} booleanProps
 * @param {Boolean} hasBreakpoint
 * @returns {String}
 */
export const generateClassNames = (props, propNames) => ''.concat(
  propNames.boolean ? getBooleanClassNames(props, propNames.boolean) : '',
  propNames.breakpoints ?
    bps.reduce((className, bp) => className.concat(
      generateBreakpointClassNames(props, propNames.breakpoints, bp)
    ), '') : ''
);

const getBooleanTypes = booleanProps => booleanProps.reduce(
  (result, prop) => Object.assign(result, { [prop]: PropTypes.bool }), {});

const getBreakpointType = bp => ({ [bp]: PropTypes.number });

const getBreakpointGenericType = (props, bp, type) => props.reduce(
  (result, prop) => Object.assign(result, { [bpPropClass(bp, prop)]: type }), {});

const generateBreakpointTypes = (propNames, bp) => Object.assign(
  propNames.include ? getBreakpointType(bp) : {},
  propNames.numeric ? getBreakpointGenericType(propNames.numeric, bp, PropTypes.number) : {},
  propNames.boolean ? getBreakpointGenericType(propNames.boolean, bp, PropTypes.bool) : {},
  propNames.visibility ? getBreakpointGenericType(propNames.visibility, bp, PropTypes.bool) : {}
);

/**
 * Generate object with prop types from props names and breakpoints
 *
 * @param {String Array} numericProps
 * @param {String Array} booleanProps
 * @param {Boolean} hasBreakpoint
 * @returns {Object}
 */
export const generatePropTypes = (propNames) =>
  Object.assign(
    propNames.boolean ? getBooleanTypes(propNames.boolean) : {},
    bps.reduce((result, bp) => Object.assign(
      result,
      generateBreakpointTypes(propNames.breakpoints, bp)
    ), {})
  );
