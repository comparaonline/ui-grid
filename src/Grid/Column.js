import React from 'react';
import cx from 'classnames';
import { generateClassNames, generatePropTypes } from './utils';

export const propNames = {
  breakpoints: {
    include: true,
    numeric: ['pull', 'push', 'offset'],
    boolean: ['centered', 'uncentered']
  }
};

export const Column = (props) => {
  const className = cx(
    props.className,
    `column${generateClassNames(props, propNames)}`
  );

  return (
    <div className={className}>{props.children}</div>
  );
};

Column.propTypes = generatePropTypes(propNames);
