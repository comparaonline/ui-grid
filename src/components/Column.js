import React from 'react';
import cx from 'classnames';
import { generateClassNames, generatePropTypes } from './utils';

export const propNames = {
  breakpoints: {
    include: true,
    numeric: ['pull', 'push', 'offset'],
    boolean: ['centered', 'uncentered'],
    visibility: ['hide', 'show']
  }
};

const Column = ({ id = null, ...props }) => {
  const className = cx(
    props.className,
    `column${generateClassNames(props, propNames)}`
  );

  const attrs = {
    className,
    ...(id && { id })
  };

  return <div {...attrs}>{props.children}</div>;
};

Column.propTypes = generatePropTypes(propNames);

export default Column;
