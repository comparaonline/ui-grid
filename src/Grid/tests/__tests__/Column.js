import React from 'react';
import { shallow } from 'enzyme';
import Column, { propNames } from '../../Column';
import { bpPropClass } from '../../utils';
import breakpoints from '../../../breakpoints.json';
import { expectClass } from '../helper';

const bps = Object.keys(breakpoints);

describe('Column', () => {
  it('renders a column', () => {
    const column = shallow(<Column />);
    expectClass(column, 'column');
    expect(column.is('div')).toBe(true);
  });

  it('renders a column with breakpoints', () => {
    bps.forEach(bp => {
      const column = shallow(<Column {...{ [bp]: 1 }} />);
      expectClass(column, `${bp}-1`);
    });
  });

  it('renders a column with numeric props', () => {
    bps.forEach(bp => {
      propNames.breakpoints.numeric.forEach(prop => {
        const column = shallow(<Column {...{ [bpPropClass(bp, prop)]: 1 }} />);
        expectClass(column, `${bp}-${prop}-1`);
      });
    });
  });

  it('renders a column with boolean props', () => {
    bps.forEach(bp => {
      propNames.breakpoints.boolean.forEach(prop => {
        const column = shallow(<Column {...{ [bpPropClass(bp, prop)]: true }} />);
        expectClass(column, `${bp}-${prop}`);
      });
    });
  });

  it('renders a row with visibility props', () => {
    Object.keys(breakpoints).forEach(bp => {
      propNames.breakpoints.visibility.forEach(prop => {
        const row = shallow(<Column {...{ [bpPropClass(bp, prop)]: true }} />);
        expectClass(row, `${prop}-for-${bp}`);
      });
    });
  });

  describe('when an additional className is pass as prop', () => {
    const className = 'test-class';
    const wrapper = shallow(<Column className={className} />);

    it('add this className', () => {
      expect(wrapper.is(`.${className}`)).toBeTruthy();
    });
  });
});
