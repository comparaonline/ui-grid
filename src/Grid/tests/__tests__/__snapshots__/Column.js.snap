// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column renders a column 1`] = `
<div
  className="column"
/>
`;

exports[`Column renders a column with boolean props 1`] = `
<div
  centeredOnSmall={true}
  className="column small-centered"
/>
`;

exports[`Column renders a column with boolean props 2`] = `
<div
  className="column small-uncentered"
  uncenteredOnSmall={true}
/>
`;

exports[`Column renders a column with boolean props 3`] = `
<div
  centeredOnMedium={true}
  className="column medium-centered"
/>
`;

exports[`Column renders a column with boolean props 4`] = `
<div
  className="column medium-uncentered"
  uncenteredOnMedium={true}
/>
`;

exports[`Column renders a column with boolean props 5`] = `
<div
  centeredOnLarge={true}
  className="column large-centered"
/>
`;

exports[`Column renders a column with boolean props 6`] = `
<div
  className="column large-uncentered"
  uncenteredOnLarge={true}
/>
`;

exports[`Column renders a column with boolean props 7`] = `
<div
  centeredOnXlarge={true}
  className="column xlarge-centered"
/>
`;

exports[`Column renders a column with boolean props 8`] = `
<div
  className="column xlarge-uncentered"
  uncenteredOnXlarge={true}
/>
`;

exports[`Column renders a column with breakpoints 1`] = `
<div
  className="column small-1"
  small={1}
/>
`;

exports[`Column renders a column with breakpoints 2`] = `
<div
  className="column medium-1"
  medium={1}
/>
`;

exports[`Column renders a column with breakpoints 3`] = `
<div
  className="column large-1"
  large={1}
/>
`;

exports[`Column renders a column with breakpoints 4`] = `
<div
  className="column xlarge-1"
  xlarge={1}
/>
`;

exports[`Column renders a column with numeric props 1`] = `
<div
  className="column small-pull-1"
  pullOnSmall={1}
/>
`;

exports[`Column renders a column with numeric props 2`] = `
<div
  className="column small-push-1"
  pushOnSmall={1}
/>
`;

exports[`Column renders a column with numeric props 3`] = `
<div
  className="column small-offset-1"
  offsetOnSmall={1}
/>
`;

exports[`Column renders a column with numeric props 4`] = `
<div
  className="column medium-pull-1"
  pullOnMedium={1}
/>
`;

exports[`Column renders a column with numeric props 5`] = `
<div
  className="column medium-push-1"
  pushOnMedium={1}
/>
`;

exports[`Column renders a column with numeric props 6`] = `
<div
  className="column medium-offset-1"
  offsetOnMedium={1}
/>
`;

exports[`Column renders a column with numeric props 7`] = `
<div
  className="column large-pull-1"
  pullOnLarge={1}
/>
`;

exports[`Column renders a column with numeric props 8`] = `
<div
  className="column large-push-1"
  pushOnLarge={1}
/>
`;

exports[`Column renders a column with numeric props 9`] = `
<div
  className="column large-offset-1"
  offsetOnLarge={1}
/>
`;

exports[`Column renders a column with numeric props 10`] = `
<div
  className="column xlarge-pull-1"
  pullOnXlarge={1}
/>
`;

exports[`Column renders a column with numeric props 11`] = `
<div
  className="column xlarge-push-1"
  pushOnXlarge={1}
/>
`;

exports[`Column renders a column with numeric props 12`] = `
<div
  className="column xlarge-offset-1"
  offsetOnXlarge={1}
/>
`;
